Dockerfile

```diff
@@ -1,40 +1,78 @@
-FROM --platform=linux/amd64 ruby:3.0.6
-RUN rm /bin/sh && ln -s /bin/bash /bin/sh
-WORKDIR /usr/src/app
-EXPOSE 3100
+ARG RUBY_VERSION=3.0.6
+FROM registry.docker.com/library/ruby:$RUBY_VERSION-slim as base
 
-RUN apt-get update -y && apt-get install -y --no-install-recommends \
-  default-jre \
-  default-mysql-client \
-  imagemagick \
-  libprotobuf-dev \
-  protobuf-compiler 
+# Rails app lives here
+WORKDIR /rails
 
-# Workaround for PhantomJS: https://github.com/DMOJ/online-judge/pull/1270
-ENV OPENSSL_CONF /etc/ssl/
+# Install base packages
+RUN apt-get update -qq && \
+    apt-get install --no-install-recommends -y \
+    curl libjemalloc2 libcurl4-openssl-dev default-libmysqlclient-dev
 
-ENV NODE_VERSION 16.18.1
-ENV NVM_DIR=/root/.nvm
+# Set production environment
+ENV RAILS_ENV="production" \
+    BUNDLE_DEPLOYMENT="1" \
+    BUNDLE_PATH="/usr/local/bundle" \
+    BUNDLE_WITHOUT="development"
 
+# Install JavaScript dependencies | Needed in base due coffee-rails gem
+ARG NODE_VERSION=16.20.2
+ARG YARN_VERSION=latest
+ENV PATH=/usr/local/node/bin:$PATH
+RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz -C /tmp/ && \
+    /tmp/node-build-master/bin/node-build "${NODE_VERSION}" /usr/local/node
+
+# Throw-away build stage to reduce size of final image
+FROM base as build
+
+# Install packages needed to build gems and node modules
+RUN apt-get install --no-install-recommends -y build-essential git libvips pkg-config
+
+# Install JavaScript dependencies
+RUN npm install -g yarn@$YARN_VERSION && rm -rf /tmp/node-build-master
+
+# Install application gems
+COPY Gemfile Gemfile.lock ./
+RUN bundle install && \
+    rm -rf ~/.bundle/ "${BUNDLE_PATH}"/ruby/*/cache "${BUNDLE_PATH}"/ruby/*/bundler/gems/*/.git && \
+    bundle exec bootsnap precompile --gemfile
+
+# Install node modules
```

Comment:
Maybe we can modify this a bit more, like this: 

# Install Node.js and Yarn
RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz -C /tmp/ && \
    /tmp/node-build-master/bin/node-build "${NODE_VERSION}" /usr/local/node && \
    npm install -g yarn@"${YARN_VERSION}" && \
    rm -rf /tmp/* /usr/local/share/.cache
    
    
 This would remove unnecessary caching by cleaning up after Node.js and Yarn installations.

Reply:
I had to split up node installation and installing the node packages since we need Node in the `base` image. I've added the line for removing the `/tmp/* and `/usr/local/share/.cache` directories in here https://github.com/GSA/search-gov/pull/1528/commits/bfba5428e3583aafe6298c26ab229737a5231a99